src/plonk/eval.rs

```diff
@@ -72,42 +73,67 @@ pub trait Eval<F: PrimeField> {
 
     /// evaluate polynomial relation on specific row
     fn eval(&self, poly: &MultiPolynomial<F>, row: usize) -> Result<F, Error> {
+        let mut evals: HashMap<(ColumnIndex, usize), F> =
+            HashMap::with_capacity(poly.degree() * poly.arity());
         let row_size = self.row_size() as i32;
         poly.monomials
             .iter()
             .map(|mono| {
                 (0..mono.arity)
-                    .map(|i| {
-                        match mono.index_to_poly[i] {
-                            // evaluation for challenge variable
-                            ColumnIndex::Challenge { column_index } => {
-                                self.eval_challenge(column_index)
-                            }
-                            // evaluation for column polynomial variable
-                            ColumnIndex::Polynominal {
-                                rotation,
-                                column_index,
-                            } => {
-                                let rotation_plus_row = rotation + (row as i32);
-                                // TODO: double check how halo2 handle
-                                // (1): row+rot < 0
-                                // (2): row+rot >= row_size = 2^K
-                                let row = if rotation_plus_row < 0 {
-                                    rotation_plus_row + row_size
-                                } else if rotation_plus_row >= row_size {
-                                    rotation_plus_row - row_size
-                                } else {
-                                    rotation_plus_row
-                                };
-
-                                self.eval_column_var(row as usize, column_index)
+                    .zip(mono.exponents.iter())
+                    .map(|(i, exp)| {
+                        if *exp == 0 {
+                            Ok(F::ONE)
+                        } else {
```

Comment:
To avoid adding additional nesting levels, please use `return`, otherwise `else` is too big
```rust
if *exp == 0 {
    return Ok(F::ONE);
}
```